lib/npg_tracking/Schema/Result/RunStatusDict.pm

```diff
@@ -122,71 +136,42 @@ our $VERSION = '0';
-
-The method croaks if no argument is supplied, if no row is found (or if,
-multiple rows are matched) otherwise the row is returned as a DBIx::Class::Row
-object.
-
-=cut
-
-sub check_row_validity {
-    my ( $self, $arg ) = @_;
+=head2 compare_to_status_description
 
-    croak 'Argument required' if !defined $arg;
+Comparison of run status descriptions based on their temporal_index values.
+Returns -1, 0 or 1 if the temporal index of run status description is less
+than, equal to or more than the temporal index for the run status description
+given as an argument.
 
-    my $field = ( $arg =~ m/^ \d+ $/msx )
-                     ? 'id_run_status_dict'
-                     : 'description' ;
-
-    my $rs = $self->result_source->schema->resultset('RunStatusDict')->
-                search( { $field => lc $arg, } );
-
-    return if $rs->count() < 1;
-
-    croak 'Panic! Multiple run_status_dict rows found' if $rs->count() > 1;
-
-    my $row = $rs->first();
-    if (!$row->iscurrent) {
-        croak 'Run status "' . $row->description . '" is not current';
-    }
-    return $row;
-}
-
-
-=head2 _insist_on_valid_row
-
-The above method is a general query tool. The user shouldn't have to deal with
-a croak just because they asked about a row that doesn't exist.
-
-This method is more severe and will croak in such a case. It is intended for
-internal methods in other classes in this library so that they don't each have
-to define their own user_validity check, but also can insist on a valid
-identifier before proceeding.
-
-It calls the above method, passing back the row object if the identifier is
-valid, but croaking if check_row_validity returns undef. If check_row_validity
-croaks (no argument, multiple rows returned) that suits this method's purpose.
-
-Could we consider caching here? To reduce the number of database queries? User
-identifiers are not likely to be created or changed during an object's
-lifetime.
+Error if the run status description given in the argument does not exist
+in the table.
 
+  print $run_status_dict_row->description(); # run_in_progress
+  print $run_status_dict_row->compare_to_status_description('analysis pending'); # -1
+  print $run_status_dict_row->compare_to_status_description('run in progress');  # 0
+  print $run_status_dict_row->compare_to_status_description('run pending');      # 1
 =cut
 
-sub _insist_on_valid_row {
-    my ( $self, $arg ) = @_;
-
-    my $row_object = $self->check_row_validity($arg);
-
-    croak "Invalid identifier: $arg" if !defined $row_object;
-
-    return $row_object;
+sub compare_to_status_description {
+  my ($self, $status_desc) = @_;
+
+  $status_desc or croak 'Non-empty status description string required';
+  my $other_status_row = $self->result_source()->resultset()
+                         ->find({description => $status_desc});
+  if (!$other_status_row) {
+    croak "Run status description '$status_desc' does not exist";
+  }
+  my $this_ti = $self->temporal_index;
```

Comment:
This can be moved earlier so we only try to find the other status row when we have checked this row has temporal index.